<a href="https://colab.research.google.com/github/FOFM030711/Caminata-aleatoria-1D/blob/main/PruebaRACHAS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <span style="color:red;">**Prueba de Rachas(Independencia)**</span>
 <span style="color:purple;"></span>

<span style="color:green;"></span>**Alumna:** Florencio Florencio Miriam Lizeth

<span style="color:brown;"></span>**Objetivo:** Realizar la prueba de independecias(Rachas) y compruebe si el ejercicio cumple con la aleatoridad de los números

A continuacion, consideremos la siguiente hipotesis nula:


$H_0$:  los numeros de la muestra ${x_i}$ son independientes

y luego

$H_1$:  los numeros de la muestra ${x_i}$ no son independientes

In [1]:
import numpy as np
from scipy.stats import norm

In [2]:
matriz = np.array([
  [0.78961, 0.76086, 0.80548, 0.58518, 0.89898, 0.28269, 0.38618, 0.79982, 0.58962,
  0.69823, 0.29931, 0.5741,  0.24,    0.93655, 0.54325, 0.58244, 0.23949, 0.19962,
  0.19147, 0.22287],
  [0.0523,  0.12079, 0.82654, 0.98611, 0.5788,  0.73059, 0.7691,  0.45679, 0.33216,
  0.17028, 0.30861, 0.26593, 0.65559, 0.88809, 0.624,   0.85853, 0.53559, 0.65002,
  0.40644, 0.07281],
 [0.10689, 0.27738, 0.29453, 0.34488, 0.67621, 0.70119, 0.68834, 0.21631, 0.03185,
  0.05475, 0.83358, 0.85903, 0.38507, 0.81772, 0.09133, 0.88752, 0.33881, 0.74579,
  0.08128, 0.64183],
 [0.55877, 0.65726, 0.20852, 0.34358, 0.0501,  0.18284, 0.5517,  0.87616, 0.61168,
  0.91512, 0.51781, 0.48338, 0.90829, 0.36982, 0.41678, 0.38729, 0.49883, 0.79113,
  0.73435, 0.44267],
 [0.14151, 0.79269, 0.42989, 0.11537, 0.00121, 0.49962, 0.1085,  0.55743, 0.09264,
  0.76262, 0.03272, 0.35288, 0.94187, 0.19904, 0.33954, 0.15506, 0.75103, 0.63453,
  0.22724, 0.72102]
])


In [3]:
matriz

array([[0.78961, 0.76086, 0.80548, 0.58518, 0.89898, 0.28269, 0.38618,
        0.79982, 0.58962, 0.69823, 0.29931, 0.5741 , 0.24   , 0.93655,
        0.54325, 0.58244, 0.23949, 0.19962, 0.19147, 0.22287],
       [0.0523 , 0.12079, 0.82654, 0.98611, 0.5788 , 0.73059, 0.7691 ,
        0.45679, 0.33216, 0.17028, 0.30861, 0.26593, 0.65559, 0.88809,
        0.624  , 0.85853, 0.53559, 0.65002, 0.40644, 0.07281],
       [0.10689, 0.27738, 0.29453, 0.34488, 0.67621, 0.70119, 0.68834,
        0.21631, 0.03185, 0.05475, 0.83358, 0.85903, 0.38507, 0.81772,
        0.09133, 0.88752, 0.33881, 0.74579, 0.08128, 0.64183],
       [0.55877, 0.65726, 0.20852, 0.34358, 0.0501 , 0.18284, 0.5517 ,
        0.87616, 0.61168, 0.91512, 0.51781, 0.48338, 0.90829, 0.36982,
        0.41678, 0.38729, 0.49883, 0.79113, 0.73435, 0.44267],
       [0.14151, 0.79269, 0.42989, 0.11537, 0.00121, 0.49962, 0.1085 ,
        0.55743, 0.09264, 0.76262, 0.03272, 0.35288, 0.94187, 0.19904,
        0.33954, 0.15506, 0.75103, 0.6

In [4]:

# Convertimos la matriz en vector
datos = matriz.flatten()
n = len(datos)


In [5]:

print("Cantidad total de datos (n):", n)


Cantidad total de datos (n): 100


<span style="color:orange;"></span>**Paso 1**: Determine la secuencia $S$ de $0$ y $1$ la + y -, así:



*   Si $x_i  \leq x_{i-1}$n escribir $0$ (o "-")
*   Si $x_i  > x_{i-1}$ escribir $0$ (o "+")



In [6]:
# Construcción de la secuencia S


secuencia = []

for i in range(1, n):
    if datos[i] <= datos[i-1]:
        secuencia.append(0)
    else:
        secuencia.append(1)




In [7]:
print("\nSecuencia generada:")
print(secuencia)


Secuencia generada:
[0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1]


La secuencia contiene $n-1$ numeros



<span style="color:orange;"></span>**Paso 2**: Contar el numero de rachas observadas $c_o$
Una racha viene dada por el numero de ceros o unas secuencias que la forman.


In [8]:

rachas = 1

for i in range(1, len(secuencia)):
    if secuencia[i] != secuencia[i-1]:
        rachas += 1

print("\nNúmero de rachas observadas (c_o):", rachas)


Número de rachas observadas (c_o): 70



<span style="color:brown;"></span>**Paso 3**: Calcule el valor esperado y la varianza del numero de rachas, asi como el estadistico $Z_0$ mediante:



In [9]:

mu = (2*n - 1) / 3
sigma = (16*n - 29) / 90   # ← SIN RAÍZ

Z0 = (rachas - mu) / sigma





$\mu_{c_0} = \frac{2n - 1}{3}$


In [10]:
print("\nValor esperado (μ):", round(mu,4))



Valor esperado (μ): 66.3333



$\sigma_{c_0}$=$\frac{16n-29}{90}$



In [11]:

print("Sigma (σ):", round(sigma,4))


Sigma (σ): 17.4556


$Z_0$=$\frac{c_o-\mu}{\sigma}$


Dado que $C_o$ sigue una distribución normal de media $μ_{c_o}$ y varianza $σ^2_{C_o}$ y $Z_0$ una distribución normal (0,1). El intervalo de aceptación:
$$
-≤Z_0\le Z_{\frac{α}{2}} $$
o bien
$$|Z_0| ≤Z_{\frac{α}{2}}$$

In [12]:

print("Estadístico Z0:", round(Z0,4))

Estadístico Z0: 0.2101


In [13]:

alpha = 0.05
Z_critico = norm.ppf(1 - alpha/2)



In [18]:

print("\nValor crítico ±Zα/2:", round(Z_critico,4))

print("\n================ RESULTADO FINAL ================")

if abs(Z0) < Z_critico:
    print("NO se rechaza H0")
    print("Conclusión: La secuencia PUEDE considerarse ALEATORIA.")
else:
    print("Se rechaza H0")
    print("Conclusión: La secuencia NO es aleatoria.")

print("=================================================")


Valor crítico ±Zα/2: 1.96

================ RESULTADO FINAL ================
NO se rechaza H0
Conclusión: La secuencia PUEDE considerarse ALEATORIA.


<span style="color:blue;"></span>**Interpretación:**

El estadístico $Z$ calculado (-0.0827) cae dentro del intervalo crítico $[-1.96, 1.96]$, lo que indica que no hay evidencia suficiente para rechazar la hipótesis de aleatoriedad.

<span style="color:purple;">Por lo tanto no se rechaza la hipotesis
$H_0:$ Los números de la muestra $\{x_i\}$ son independientes(aleatorios)</span>


<span style="color:pink;"></span>***Conclusión***

El programa desarrollado para la Prueba de Independencia (Prueba de Rachas)
nos permite analizar de manera clara y estructurada si una secuencia de números
realmente se comporta como debería hacerlo un conjunto de datos aleatorios.
Más allá de realizar cálculos, este algoritmo nos ayuda a comprender el
comportamiento interno de los datos y a detectar posibles patrones ocultos
que podrían afectar nuestros resultados.

En este ejemplo, con un nivel de significancia del $5\%$, los resultados nos permiten tener confianza en el comportamiento de la secuencia analizada. Al leer los números por columnas, no se observan señales de algún patrón oculto o tendencia sistemática en su generación.

El número de corridas obtenido ($70$) es muy cercano al valor esperado teóricamente ($66.33$). Esta pequeña diferencia es completamente razonable y puede explicarse por la variabilidad natural que siempre está presente cuando se trabaja con datos aleatorios. Es decir, las ligeras discrepancias observadas son consistentes con lo que normalmente ocurre por simple azar.

En términos prácticos, esto indica que la secuencia cumple adecuadamente con la propiedad de independencia, por lo que puede emplearse con confianza en simulaciones y en métodos probabilísticos donde la aleatoriedad y la ausencia de patrones sistemáticos son fundamentales.
